### DISCUSSION TITLE
Maintaining Multiple Relation Tables

### DISCUSSION THREAD

**Discussion by Sand-NGR** [pocketbase/pocketbase#3788] [Q&A] [answered]
Hi. I have two collections like this:

1- Posts: Has a column (`post_comments`) that is a multiple relation to a `comments` table.
2- Comments: Has a column (`post_id`) that is a single relation to a `posts` Table.

Let's say a user submits a comment on a post. How do I update both collections with one query? The comment should be submitted to the `comments` collection and the `post_id` will be set fine but the `post_comments` column on the `posts` collection needs to be updated as well to include the new comment. Is that a separate query that I have to do with the pocketbase SDK? Is this a bad design on my part?

**Comment by ganigeorgiev** [comment, marked as answer]
There is no builtin automatic way to "sync" relations across collections at the moment, but if you are using the JS pb_hooks, you can attach to the [onRecordBeforeCreateRequest](https://pocketbase.io/docs/js-event-hooks#onrecordbeforecreaterequest) and [onRecordBeforeUpdateRequest](https://pocketbase.io/docs/js-event-hooks#onrecordbeforeupdaterequest) and perform the update queries manually but you'll have to do that for each collection.

> Is this a bad design on my part?

I guess it depends, but I personally would try to avoid having to maintain the relation in multiple places and instead will try to keep only the `comments.post_id` relation.

If the only reason you want to maintain the relation in multiple places is for expand, note that you can also use [back-relation expansion](https://pocketbase.io/docs/working-with-relations/).

**Reply by Sand-NGR** [reply]
do you have a couple of examples for these pb_hooks? examples from the docs are all console.log statements.

**Reply by ganigeorgiev** [reply]
@Sand-NGR Some elaboration on why do you need to maintain the relation in multiple collections in the first place will be helpful to understand the use case and maybe provide a better alternative.

I agree that the docs aren't great at the moment, but examples for the hooks are difficult to generalize and if you expect a one line solution then you'll not find here.
If you really want to go with the hooks approach here is an example how you could update the `posts.post_comments` relation when a new comment is created (_assuming that a `comments.post_id` doesn't change, aka. it has an update rule like: `@request.data.post_id:isset = false || @request.data.post_id = post_id`_):

```js
// pb_hooks/main.pb.js

// docs reference:
// - https://pocketbase.io/docs/js-event-hooks/#record-crud-api-hooks
// - https://pocketbase.io/docs/js-records/
onRecordBeforeCreateRequest((e) => {
    // retrieve the comment's post
    const post = $app.dao().findRecordById("posts", e.record.get("post_id"));

    // add the comment id to the post list
    const ids = post.get("post_comments");
    ids.push(e.record.id);
    post.set("post_comments", ids);

    // persist the change
    $app.dao().saveRecord(post);
}, "comments")
```

But again, without details about your use case I don't recommend the above and instead follow a structure similar to the example in https://pocketbase.io/docs/working-with-relations/.

**Reply by Sand-NGR** [reply]
I decided to use back relations. It's a better approach. I just wanted some idea of the hooks because I'm planning to use them somewhere else in my app and I didn't want to spam another discussion. Thank you, this helps a lot in understanding hooks.

**Reply by Sand-NGR** [reply]
@ganigeorgiev Is there a way to filter the `expand` field if I use back expanding? I want to get a posts comments but I show them paginated and don't retrieve all in one go. Basically I have a comments modal that retrieves comments when it's opened in a paginated manner with `getList`. Using back expanding means that I query the posts collection and get the comments through that which means I would have to filter expand.

The example in https://pocketbase.io/docs/working-with-relations/ gets all the comments at once, right?

**Reply by ganigeorgiev** [reply]
`expand` expands all relation items that can be viewed by the request client, there is no support for pagination, filtering or sorting. If you need pagination you'll have to send a separate request to the comments collection.
